File: web/src/Entity/DossierMedical.php
<?php

declare(strict_types=1);

namespace App\Entity;

use Doctrine\Common\Collections\ArrayCollection;
use Doctrine\Common\Collections\Collection;
use Doctrine\DBAL\Types\Types;
use Doctrine\ORM\Mapping as ORM;
use Symfony\Component\Validator\Constraints as Assert;

#[ORM\Entity(repositoryClass: \App\Repository\DossierMedicalRepository::class)]
#[ORM\Table(name: 'dossier_medical')]
#[ORM\HasLifecycleCallbacks]
class DossierMedical
{
    #[ORM\Id]
    #[ORM\GeneratedValue]
    #[ORM\Column(type: Types::INTEGER)]
    private ?int $id = null;

    #[ORM\Column(type: Types::STRING, length: 50, unique: true)]
    #[Assert\NotBlank(message: 'Le numéro de dossier est obligatoire.')]
    #[Assert\Length(min: 2, max: 50, minMessage: 'Le numéro doit contenir au moins {{ limit }} caractères.', maxMessage: 'Le numéro ne peut pas dépasser {{ limit }} caractères.')]
    private ?string $numeroDossier = null;

    #[ORM\Column(type: Types::STRING, length: 120)]
    #[Assert\NotBlank(message: 'Le nom du patient est obligatoire.')]
    #[Assert\Length(min: 3, max: 120, minMessage: 'Le nom doit contenir au moins {{ limit }} caractères.', maxMessage: 'Le nom ne peut pas dépasser {{ limit }} caractères.')]
    private ?string $patientNom = null;

    #[ORM\Column(type: Types::STRING, length: 120)]
    #[Assert\NotBlank(message: 'Le prénom du patient est obligatoire.')]
    #[Assert\Length(min: 3, max: 120, minMessage: 'Le prénom doit contenir au moins {{ limit }} caractères.', maxMessage: 'Le prénom ne peut pas dépasser {{ limit }} caractères.')]
    private ?string $patientPrenom = null;

    #[ORM\Column(type: Types::DATE_MUTABLE, nullable: true)]
    #[Assert\NotNull(message: 'La date de naissance est obligatoire.')]
    #[Assert\LessThan('today', message: 'La date de naissance doit être dans le passé.')]
    private ?\DateTimeInterface $dateNaissance = null;

    #[ORM\Column(type: Types::STRING, length: 20, nullable: true)]
    #[Assert\Choice(choices: ['M', 'F', 'Autre'], message: 'Le genre doit être M, F ou Autre.')]
    private ?string $genre = null;

    #[ORM\Column(type: Types::STRING, length: 180, nullable: true)]
    #[Assert\Email(message: 'L\'adresse email "{{ value }}" n\'est pas valide.')]
    #[Assert\Length(max: 180)]
    private ?string $email = null;

    #[ORM\Column(type: Types::STRING, length: 30, nullable: true)]
    #[Assert\Regex(pattern: '/^[\d\s\+\-\(\)]{8,30}$/', message: 'Le numéro de téléphone n\'est pas valide (chiffres, espaces, +, -, parenthèses, 8 à 30 caractères).')]
    private ?string $telephone = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    #[Assert\Length(max: 2000, maxMessage: 'L\'adresse ne peut pas dépasser {{ limit }} caractères.')]
    private ?string $adresse = null;

    #[ORM\Column(type: Types::TEXT, nullable: true)]
    #[Assert\Length(max: 5000, maxMessage: 'Les remarques ne peuvent pas dépasser {{ limit }} caractères.')]
    private ?string $remarques = null;

    #[ORM\Column(type: Types::DATETIME_IMMUTABLE)]
    private ?\DateTimeImmutable $dateCreation = null;

    #[ORM\Column(type: Types::DATETIME_MUTABLE, nullable: true)]
    private ?\DateTimeInterface $dateModification = null;

    /** @var Collection<int, Document> */
    #[ORM\OneToMany(targetEntity: Document::class, mappedBy: 'dossierMedical', cascade: ['persist', 'remove'], orphanRemoval: true)]
    #[ORM\OrderBy(['dateDocument' => 'DESC', 'dateCreation' => 'DESC'])]
    private Collection $documents;

    public function __construct()
    {
        $this->documents = new ArrayCollection();
    }

    #[ORM\PrePersist]
    public function setDateCreationValue(): void
    {
        if ($this->dateCreation === null) {
            $this->dateCreation = new \DateTimeImmutable();
        }
    }

    #[ORM\PreUpdate]
    public function setDateModificationValue(): void
    {
        $this->dateModification = new \DateTime();
    }

    public function getId(): ?int
    {
        return $this->id;
    }

    public function getNumeroDossier(): ?string
    {
        return $this->numeroDossier;
    }

    public function setNumeroDossier(string $numeroDossier): static
    {
        $this->numeroDossier = $numeroDossier;
        return $this;
    }

    public function getPatientNom(): ?string
    {
        return $this->patientNom;
    }

    public function setPatientNom(string $patientNom): static
    {
        $this->patientNom = $patientNom;
        return $this;
    }

    public function getPatientPrenom(): ?string
    {
        return $this->patientPrenom;
    }

    public function setPatientPrenom(string $patientPrenom): static
    {
        $this->patientPrenom = $patientPrenom;
        return $this;
    }

    public function getDateNaissance(): ?\DateTimeInterface
    {
        return $this->dateNaissance;
    }

    public function setDateNaissance(?\DateTimeInterface $dateNaissance): static
    {
        $this->dateNaissance = $dateNaissance;
        return $this;
    }

    public function getGenre(): ?string
    {
        return $this->genre;
    }

    public function setGenre(?string $genre): static
    {
        $this->genre = $genre;
        return $this;
    }

    public function getEmail(): ?string
    {
        return $this->email;
    }

    public function setEmail(?string $email): static
    {
        $this->email = $email;
        return $this;
    }

    public function getTelephone(): ?string
    {
        return $this->telephone;
    }

    public function setTelephone(?string $telephone): static
    {
        $this->telephone = $telephone;
        return $this;
    }

    public function getAdresse(): ?string
    {
        return $this->adresse;
    }

    public function setAdresse(?string $adresse): static
    {
        $this->adresse = $adresse;
        return $this;
    }

    public function getRemarques(): ?string
    {
        return $this->remarques;
    }

    public function setRemarques(?string $remarques): static
    {
        $this->remarques = $remarques;
        return $this;
    }

    public function getDateCreation(): ?\DateTimeImmutable
    {
        return $this->dateCreation;
    }

    public function getDateModification(): ?\DateTimeInterface
    {
        return $this->dateModification;
    }

    /** @return Collection<int, Document> */
    public function getDocuments(): Collection
    {
        return $this->documents;
    }

    public function addDocument(Document $document): static
    {
        if (!$this->documents->contains($document)) {
            $this->documents->add($document);
            $document->setDossierMedical($this);
        }
        return $this;
    }

    public function removeDocument(Document $document): static
    {
        if ($this->documents->removeElement($document)) {
            if ($document->getDossierMedical() === $this) {
                $document->setDossierMedical(null);
            }
        }
        return $this;
    }

    public function __toString(): string
    {
        return sprintf('%s - %s %s', $this->numeroDossier ?? '', $this->patientNom ?? '', $this->patientPrenom ?? '');
    }
}

File: javafx-app/src/main/java/org/docbook/entities/records/DossierMedical.java
package org.docbook.entities.records;

import java.time.LocalDate;
import java.time.LocalDateTime;

public class DossierMedical {

    private int id;
    private String numeroDossier;
    private String patientNom;
    private String patientPrenom;
    private LocalDate dateNaissance;
    private String genre;
    private String email;
    private String telephone;
    private String adresse;
    private String remarques;
    private LocalDateTime dateCreation;
    private LocalDateTime dateModification;

    public DossierMedical() {
    }

    public DossierMedical(int id, String numeroDossier, String patientNom, String patientPrenom, LocalDate dateNaissance, String genre, String email, String telephone, String adresse, String remarques, LocalDateTime dateCreation, LocalDateTime dateModification) {
        this.id = id;
        this.numeroDossier = numeroDossier;
        this.patientNom = patientNom;
        this.patientPrenom = patientPrenom;
        this.dateNaissance = dateNaissance;
        this.genre = genre;
        this.email = email;
        this.telephone = telephone;
        this.adresse = adresse;
        this.remarques = remarques;
        this.dateCreation = dateCreation;
        this.dateModification = dateModification;
    }

    // Getters and Setters
    public int getId() { return id; }
    public void setId(int id) { this.id = id; }

    public String getNumeroDossier() { return numeroDossier; }
    public void setNumeroDossier(String numeroDossier) { this.numeroDossier = numeroDossier; }

    public String getPatientNom() { return patientNom; }
    public void setPatientNom(String patientNom) { this.patientNom = patientNom; }

    public String getPatientPrenom() { return patientPrenom; }
    public void setPatientPrenom(String patientPrenom) { this.patientPrenom = patientPrenom; }

    public LocalDate getDateNaissance() { return dateNaissance; }
    public void setDateNaissance(LocalDate dateNaissance) { this.dateNaissance = dateNaissance; }

    public String getGenre() { return genre; }
    public void setGenre(String genre) { this.genre = genre; }

    public String getEmail() { return email; }
    public void setEmail(String email) { this.email = email; }

    public String getTelephone() { return telephone; }
    public void setTelephone(String telephone) { this.telephone = telephone; }

    public String getAdresse() { return adresse; }
    public void setAdresse(String adresse) { this.adresse = adresse; }

    public String getRemarques() { return remarques; }
    public void setRemarques(String remarques) { this.remarques = remarques; }

    public LocalDateTime getDateCreation() { return dateCreation; }
    public void setDateCreation(LocalDateTime dateCreation) { this.dateCreation = dateCreation; }

    public LocalDateTime getDateModification() { return dateModification; }
    public void setDateModification(LocalDateTime dateModification) { this.dateModification = dateModification; }

    @Override
    public String toString() {
        return "DossierMedical{" +
                "id=" + id +
                ", numeroDossier='" + numeroDossier + '\'' +
                ", patientNom='" + patientNom + '\'' +
                ", patientPrenom='" + patientPrenom + '\'' +
                '}';
    }
}


